### DISCUSSION TITLE
Environment variables at build time

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#67] [General]
Hi guys,
right now handle the environment variables of the frontend services is pretty tricky; someone use the backend `/config` API, someone use a DB and so on.

Is planned to handle the environment variables at build time too?
I don't really know if I agree with this proposal, because it's pretty dangerous to hardly link a container image to a specific environment.
Moreover the Mia-Platform console right now doesn't directly handle the build of the images.

However there are frontend frameworks that handle the environment variables in this way, for example Angular have the environment files (`variables-{{ENVIRONMENT}}.json`).

What do you think about that?

**Comment by davidebianchi** [comment]
To cover this use case, we probably should change how we handle frontend. By the way, it is a decision of the team working on the project, from my point of view I don't like to make build per environment. 
A possible solution could be to inject in the container of the website a file at deploy time, which will be served directly from the container.

The approach I currently prefer is similar to this, and handle the feature by feature toggle. And instead to create a docker image  for each frontend, to have a static file platform which can also handle the case.

I suggest you also to see the [OpenFeature](https://docs.openfeature.dev/docs/reference/intro/) spec for the feature toggle approach.

**Comment by davidetan** [comment]
FE developers are used to features offered by their preferred build tools.
Working with Mia-Platform Console should not prevent them on using such features.

Regarding custom env variables feature for FEs, for example, we have:

- [CRA](https://create-react-app.dev/docs/adding-custom-environment-variables/#referencing-environment-variables-in-the-html) that enables the usage of env variable with prefix `REACT_APP` 
- [Vite](https://vitejs.dev/guide/env-and-mode.html) that does the same but using the prefix `VITE_`
- [NextJS](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) uses the previx `NEXT_PUBLIC_`

Platforms that enable the usage of these tools/frameworks/... usually support such a feature.

Some example are Vercel and SST.

- SST explicitly supports and documents the three I listed above:
  - [CRA](https://docs.sst.dev/constructs/ReactStaticSite#environment-variables)
  - [Vite](https://docs.sst.dev/constructs/ViteStaticSite#environment-variables)
  - [NextJS](https://docs.sst.dev/constructs/NextjsSite#environment-variables), though here there are some caveats
- Vercel has a great doc showing that, and also [provide a list](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) of additional env variables provided depending on the chosen tool.

**Reply by davidebianchi** [reply]
Yes, some framework add the feature but it's a choice of the developer if use it or not. The Console does not actually prevent you from use any of the features you have listed.

**Reply by davidetan** [reply]
~~So, I did not get the question.~~
Ok, I got the question. But saying:

> The Console does not actually prevent you from use any of the features you have listed.

How can a Console user use the feature listed above?

**Reply by davidebianchi** [reply]
As I say above, it depends by the use case. 
If we take the create react app approach, for example, you should create one build per environment. So, the approach would be to create three different build, with the proper env var injected at build time. All of this can be made from the pipeline build, and from Console you can address to different image tags by environment (interpolating with public variables). I don't recommend this approach, but developer are free to use it.

For example, for the development of the Console itself, we use [staticdeploy](https://staticdeploy.io/) managed by a Console project to create self hosted PR previews of the static files, which support the inject of env variables in another way.

So, we could better integrate frontend development to simplify it to the end user in the future, and if I correctly remember it is already in the roadmap future ideas. But Console should not limit the choice. If you have some trouble using one of the following approaches, I can support you (maybe Console has some limitation I don't know which we can resolve, I haven't tested all the possibilities listed in your response).

**Reply by davidetan** [reply]
I agree the Console should not limit the choice, but at the same time it should not make extremely difficult using such a widespread feature.

Until there won't be a straightforward way to enable custom env var in FE without using a CRUD or a custom service there should be at least a discussion (like this one), a tutorial or another resource explaining how to use custom env var within your FE.
Otherwise, if we discourage the usage of such a feature, we still should have a related page explaining why and the alternatives we propose.

**Reply by davidebianchi** [reply]
I agree with you, it would be very interesting a tutorial/blog post on the theme. We should add it describing different solutions, pros and cons! We should add in the pipeline of the documentation part, meanwhile I think that in this discussion there are a lot of ideas to start with. 😊

If you need more help about the best solution for your use case we can talk about it more in detail!

As a note: this is a common problem also for the community, for example it's possible to see [this issue of CRA](https://github.com/facebook/create-react-app/issues/2353).

**Comment by giulioroggero** [comment]
Nice conversation. What do you think about @akelity to add a guide in https://docs.mia-platform.eu/ for that kind of problems? @daaaaa @davidebianchi @silversoul93 your contributions are welcome!

**Reply by frengooo** [reply]
I think it's a great idea. Probably for users that frequently develop frontend with the Console is pretty easy to know how to handle the static files and environment variables. For other users is not so simple, a guide would be great.
@akelity I'm happy to help you 😄
